src/analysis/cache.Rmd

```diff
@@ -0,0 +1,313 @@
+
+For each stat and each opcode we calculate
+
+```
+slope(opcode)/slope(EXP)
+```
+
+where dividing by slope(EXP) is for the normalization purposes. We could think of this as the average cache usage per opcode compared to EXP opcode.
+
+```{r fig.width=10}
+a <- split(t, t$opcode)
+ns <- c("task_clock", "instructions", "L1_dcache_loads", "L1_icache_loads", "dTLB_loads", "iTLB_loads")
+slopes <- list()
+for (j in 1:length(a)) {
+  n <- names(a)[j]
+  i <- a[j][[n]]
+  s <- lapply(paste(ns, '~ op_count'), function(f){
+    b <- lm(as.formula(f), data=i)
+    b$coefficients["op_count"]
+  })
+  if (n == "EXP") {
+    EXPslopes <- s
+  }
+  slopes <- append(slopes, s)
+}
+for (j in 1:length(slopes)) {
+  slopes[[j]] = slopes[[j]] / EXPslopes[[((j-1) %% length(ns)) + 1]]
+}
+slopes <- matrix(slopes, nrow = length(ns), ncol = length(a), byrow = FALSE)
+matplot(slopes, type="l", xlab = "stats", axes=FALSE)
+axis(side=1, at=1:length(ns), labels=ns)
+axis(2)
+```
+
+### L1_icache_load_misses, dTLB_load_misses, iTLB_load_misses
+
+Note that L1_dcache_load_misses are not accounted, there is no data. 
+
+The number of misses indicates how often a given item of data was accessed but not present in the cache.
+
+To have better overview, we calculate maximal miss ratio for each opcode. See the table below.
+
+```{r}
+sqldf("SELECT opcode, max(L1_icache_load_misses/L1_icache_loads) as 'max L1_ichache miss ratio', max(dTLB_load_misses/dTLB_loads) as 'max dTLB miss ratio', max(iTLB_load_misses/iTLB_loads) as 'max iTLB miss ratio' FROM measurements_avg GROUP BY opcode")
+```
+
+The miss ratio is 0.1% at most for L1 instr cache. This indicates that executions tightly depends on L1 caches. Probably the required evmone executable fits in L1 cache. Unfortunately, stats for L1 data cache is missing. Note that a contract bytecode is taken from L1 data cache.
+
+The miss ratios for translation tables are extremely low. This is expected. The most important is that this does not affect comparison of opcode workloads.
+
+### LLC_loads, LLC_load_misses
+
+The miss ratio for the last level cache is very low, 0.02% at most. See the table below.
+
+```{r}
+sqldf("SELECT opcode, max(LLC_load_misses/LLC_loads) as 'max LLC miss ratio' FROM measurements_avg GROUP BY opcode")
+```
+
+One observation is that our tests hardly leave caches. The other observation is that every opcode has almost the same ratio, which is very good.
+
+Finally, we compare usage of L1 caches and LLC caches. The formula is `LLC_loads/(L1_icache_loads+L1_dcache_loads)`
+
+```{r}
+sqldf("SELECT opcode, avg(LLC_loads/(L1_icache_loads+L1_dcache_loads)) as ratio, stdev(LLC_loads/(L1_icache_loads+L1_dcache_loads)) as sdev FROM measurements_avg GROUP BY opcode")
+```
+
+Every opcode has similar ratio. L1 cache is 1000 times more often in use than LLC. So L1 cache performance has crucial impact on test results.
```

Comment:
EXP looks significantly lower. We could use some model to make a solid argument if it is statistically different.

For now let's continue examining "by eye", and I'll make a self note to propose some model